File: backend/src/main/java/br/com/calciolari/datahub/imports/infrastructure/interpdv/qrp/BrazilianDecimalParser.java
package br.com.calciolari.datahub.imports.infrastructure.interpdv.qrp;

import java.math.BigDecimal;
import java.util.regex.Pattern;

/**
 * Locale-tolerant decimal parser ported from the PoC {@code brNumber} helper
 * ({@code docs/poc/index.html}), returning {@link BigDecimal} instead of IEEE-754.
 */
public final class BrazilianDecimalParser {

	private static final Pattern CURRENCY_PREFIX = Pattern.compile("(?i)R\\$\\s*");
	private static final Pattern NON_NUMERIC = Pattern.compile("[^0-9+\\-.]");

	private BrazilianDecimalParser() {
	}

	public static BigDecimal parse(String raw) {
		if (raw == null || raw.isBlank()) {
			return null;
		}
		String v = CURRENCY_PREFIX.matcher(raw.trim()).replaceFirst("").trim();
		if (v.indexOf(',') >= 0 && v.indexOf('.') >= 0) {
			v = v.replace(".", "").replace(',', '.');
		}
		else if (v.indexOf(',') >= 0) {
			v = v.replace(',', '.');
		}
		v = NON_NUMERIC.matcher(v).replaceAll("");
		if (v.isEmpty() || "-".equals(v) || "+".equals(v) || ".".equals(v)) {
			return null;
		}
		try {
			return new BigDecimal(v);
		}
		catch (NumberFormatException ex) {
			return null;
		}
	}
}

File: backend/src/Calciolari.DataHub/Imports/Infrastructure/InterPdv/Qrp/BrazilianDecimalParser.cs
using System.Globalization;
using System.Text.RegularExpressions;

namespace Calciolari.DataHub.Imports.Infrastructure.InterPdv.Qrp;

/// <summary>
/// Locale-tolerant decimal parser ported from the PoC brNumber helper,
/// returning decimal instead of IEEE-754.
/// </summary>
public static class BrazilianDecimalParser
{
    private static readonly Regex CurrencyPrefix = new(@"R\$\s*", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex NonNumeric = new(@"[^0-9+\-.]", RegexOptions.Compiled);

    public static decimal? Parse(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw))
        {
            return null;
        }

        var v = CurrencyPrefix.Replace(raw.Trim(), string.Empty, 1).Trim();
        if (v.Contains(',') && v.Contains('.'))
        {
            v = v.Replace(".", string.Empty).Replace(',', '.');
        }
        else if (v.Contains(','))
        {
            v = v.Replace(',', '.');
        }

        v = NonNumeric.Replace(v, string.Empty);
        if (v.Length == 0 || v is "-" or "+" or ".")
        {
            return null;
        }

        return decimal.TryParse(v, NumberStyles.Number, CultureInfo.InvariantCulture, out var parsed)
            ? parsed
            : null;
    }
}
